Add nextjs-adapter api docs to readme #89 (#109)
enonic/nextjs-adapter: README.md

> NextJS adapter with Guillotine support and basic views

## Installation

---

```bash
npm i --save @enonic/nextjs-adapter
```

## Usage

---

Just import required functions and views in your project:

```tsx
// all functions and interfaces are accessible at the library root
import {fetchContent} from "@enonic/nextjs-adapter";

> // views are placed in a folder called views and are default exports in files

Reviewer: ```suggestion
// views are placed in a folder called `views` and are default exports in files
```

import StaticContent from "@enonic/nextjs-adapter/views/StaticContent"

// ...
const {page} = await fetchContent(path, context);
// ...

return <StaticContent/>
```

## API

---

### Functions

They are available at the library root at `@enonic/nextjs-adapter`

### <a id="fetch-content"></a>`fetchContent(contentPath: string | string[], context: Context) => Promise<FetchContentResult>`

> Fetches the content/component by path with component queries optimization, page structure as well as runtime info calculation. This is the

Reviewer: Unnecessary linebreak?

Author: Doesn't affect the output, but easier to navigate the source

        message: string
    } | null;
    data: Record<string, any> | null,       // Result of the content query
    common: Record<string, any> | null,     // Result of the common query
    meta: MetaData,                         // Runtime information
    page: PageComponent | null,             // The structure of the page (not present when rendering single component)
};
```

---

### <a id="fetch-guillotine"></a>`fetchGuillotine(apiUrl: string, body: ContentApiBaseBody, headers?: {}) => Promise<GuillotineResult>`

Makes custom guillotine request. Used by [fetchContent()](#fetch-content) method.

| Argument              | Description                                 |
|-----------------------|---------------------------------------------|
| `apiUrl`              | Guillotine API URL                          |
| `body`                | Body of the request                         |
| `headers = undefined` | Object of custom headers with string values |

Usage:

```tsx
import {fetchGuillotine} from '@enonic/nextjs-adapter';

const body = {
    query: 'qraphql query as string',
    variables: {
        foo: 'bar',
    },
};

const headers = {
    'custom-header': 'header-value',
};

> const result = fetchFromApi('http://domain:8000/graphql/api', body, contentPath, headers,);

Reviewer: `contentPath` is used here as an argument but not described in the table of arguments above.

Usage:

```tsx
import {fetchFromApi} from '@enonic/nextjs-adapter';

const body = {
    query: 'qraphql query as string',
    variables: {
        foo: 'bar',
    },
}

const headers = {
    'custom-header': 'header-value',
}

const result = fetchFromApi('http://domain:8000/graphql/api', body, headers, 'GET');
```

---

### `getUrl(url: string, meta: MetaData) => string`

Converts an absolute URL to the relative one for current viewer (Next.js/Enonic XP).

> **INFO:** For your URLs to work both in Enonic XP and Next.js you need to:
> 1. Query absolute URLs from guillotine
> 2. Wrap them with `getUrl()` function in the views

| Argument | Description                                             |
|----------|---------------------------------------------------------|
| `url`    | URL you want to transform                               |
| `meta`   | Runtime data returned by [fetchContent](#fetch-content) |

Usage:

```tsx
import {getUrl} from '@enonic/nextjs-adapter';

const urlRelativeToViewer = getUrl('http://domain:8000/graphql/api', meta);
```

---

### <a id="rich-text-query"></a>`richTextQuery(fieldName: string) => string`

> This is a utility function for querying for `RichTextData` needed for [RichTextView](#rich-text-view). It creates a graphql query string for

Reviewer: Unnecessary linebreak?

Author: Doesn't affect the output, but easier to navigate the source

HTML area input type with given field name.

| Argument    | Description          |
|-------------|----------------------|
| `fieldName` | HTML area field name |

Usage:

```tsx
import {richTextQuery} from '@enonic/nextjs-adapter';

const query = `query($path:ID!){
    guillotine {
        get(key:$path) {
            _path
            type
            ${richTextQuery('htmlField')}
        }
    }
}`;
```

### `ComponentRegistry`

> Registry containing definitions of all components (i.e. pages, parts, layouts, macros, etc. ). It is used in the runtime by `nextjs-adapter`

Reviewer: Unnecessary linebreak?

Author: Doesn't affect the output, but easier to navigate the source

to make component queries and render components. It has several public methods:

---

#### `setCommonQuery(query: SelectedQueryMaybeVariablesFunc): void`

> Setups the common query that is going to be executed along with component queries and passed to every component on the page.

Reviewer: ```suggestion
Sets up a common query that is going to be executed along with component queries and passed to every component on the page.
```

| Argument | Description                           |
|----------|---------------------------------------|
| `query`  | Common [query definition](#query-def) |

Usage:

```tsx
import {ComponentRegistry} from '@enonic/nextjs-adapter';

const query = {
    query: 'graphql query',
    variables: (path, context, config) => {
        return {
            path: path + '/some/processing'
        };
    }
}

ComponentRegistry.setCommonQuery(query);
```

---

#### `getCommonQuery(): SelectedQueryMaybeVariablesFunc`

Gets the common query definition.

<a id="query-def"></a>Response type:

```tsx
type SelectedQueryMaybeVariablesFunc =
    string |
    QueryGetter |
    {
        query: string | QueryGetter,
        variables: VariablesGetter
    } |
    [string | QueryGetter, VariablesGetter];
```

Usage:

```tsx
import {ComponentRegistry} from '@enonic/nextjs-adapter';

const query = ComponentRegistry.getCommonQuery();
```

---

#### `getByComponent(component: PageComponent): ComponentDefinition | undefined`

> Gets the definition of component stored in the ComponentRegistry by component.

Reviewer: ```suggestion
Gets component definition from the ComponentRegistry.
```

Gets the component definition stored in ComponentRegistry by its name.

> **NOTE:** Read [addComponent](#add-comp) note before using `getComponent`.

| Argument | Description    |
|----------|----------------|
| `name`   | Component name |

Usage:

```tsx
import {ComponentRegistry} from '@enonic/nextjs-adapter';

const definition = ComponentRegistry.getMacro('macro-name');
```

Response type: [component definition](#comp-def)

---

#### `getMacros(): ComponentDefinition[]`

#### `getParts(): ComponentDefinition[]`

#### `getLayouts(): ComponentDefinition[]`

#### `getPages(): ComponentDefinition[]`

#### `getContentTypes(): ComponentDefinition[]`

#### `getComponents(): ComponentDefinition[]`

Gets all component definition stored in ComponentRegistry.

> **NOTE:** Read [addComponent](#add-comp) note before using `getComponents`.

Usage:

```tsx
import {ComponentRegistry} from '@enonic/nextjs-adapter';

const macros = ComponentRegistry.getMacros();
```

Response type: List of [component definitions](#comp-def)

### `UrlProcessor`

Helper singleton for processing URLs.

---

#### <a id="process"></a>`static process(url: string, meta: MetaData): string`

> Processes the absolute URL to become relative for the current viewer, while keeping in mind Next.js assets and Enonic XP binary content

Reviewer: Unnecessary linebreak?

Author: Doesn't affect the output, but easier to navigate the source
